### ISSUE TITLE
Multi-stage Dockerfile with nvidia/cuda and Ubuntu

### ISSUE THREAD

**Issue by Alexxx77** [NVIDIA/nvidia-docker#675]
### 1. Issue or feature description
I would like to use the the provided nvidia/cuda container as part of a docker multi-stage build in order to make it easily deployable in the future, even if my main image changes etc.
You have also mentioned that this is the best approach, unfortunately I have not found an example, so that I am struggeling to write a working Dockerfile.
In this example I just use the ubuntu:16.04 image as a base image, but the actual idea is to use a more complicated image that is based on ubuntu. I know that in this way it does not make sense, because nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04 is already based on ubuntu:16.04. This is just to simplify the problem.

### 2. Steps to reproduce the issue

Dockerfile:

I have tried to reverse engineer the cudnn-devel Dockerfile [provided here](https://gitlab.com/nvidia/cuda/blob/ubuntu16.04/9.1/devel/cudnn7/Dockerfile). And include necessary commands - but it seems that I missed some ;-)

```
FROM ubuntu:16.04

COPY --from=nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04 /usr/local/cuda-9.0/ /usr/local/cuda-9.0

RUN ln -s /usr/local/cuda-9.0 /usr/local/cuda &&\
	rm -rf /var/lib/apt/lists/*

ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:${PATH}
ENV LD_LIBRARY_PATH /usr/local/cuda/lib:/usr/local/cuda/lib64
ENV LIBRARY_PATH /usr/local/cuda/lib64/stubs:${LIBRARY_PATH}

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES all
ENV NVIDIA_REQUIRE_CUDA "cuda>=9.0"
```

### 3. Information
Host:
Ubuntu 16.04
Docker version 17.12.1-ce
NVIDIA-SMI 384.111
CUDA 9.0 ( V9.0.176) / cuDNN 7.0

The error is that when I am for example try to run the YOLO9000 network ([darkflow](https://github.com/thtrieu/darkflow)) that the libcudnn.so.7 is missing. And that makes sense because on the original 9.0-cudnn7-devel-ubuntu16.04  container that file is located under `/usr/lib/x86_64-linux-gnu/libcudnn.so.7` and I did not copy it.

Basically my problem is that I do not know what files all need to be copied to the final Docker.
Please advise. I guess such an example would also be extremely helpful for other people.
@flx42 : The only thing that I have found was a [multi-stage deviceQuery](https://gist.github.com/flx42/91a501dcd7f64a0845808b99f2a00d06) from you. Maybe you have already found a solution to create a multi-stage build Dockerfile with nvidia/cuda?
Or do you recommend to just create a Dockerfile that installs CUDA again inside of the container?

Thank you.

Regards
Alex

**Comment by flx42** [comment]
You have the wrong approach, take a look at this example:
https://gitlab.com/nvidia/samples/blob/master/cuda/ubuntu16.04/ffmpeg-gpu/Dockerfile

In the first stage, you can install `gcc`, `nvcc`, all the CUDA headers, etc. Those are the **build-time** dependencies. Then you compile your app.

In the second stage, you should only copy the binaries/libraries **from your app**. Ideally, you could even generate a deb/rpm/pip package and you will pass that down to the next stage.
In this second stage, either you do `FROM nvidia/cuda:9.0-cudnn7-runtime`  and then you have all the CUDA libraries necessary to run your app (including `libcudnn.so.7`).
Or, you do `FROM nvidia/cuda:9.0-base` but then you will have to manually install your **runtime dependencies**. TensorFlow does something similar now, but not with a multi-stage build, the pip package is downloaded instead.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/docker/Dockerfile.gpu

**Comment by Alexxx77** [comment]
Thank you Felix for the quick answer.
It completely makes sense to bring the binaries from the app to the cuda image and not the other way around.
In my specific use case this is unfortunately not that easy. What I am trying to achieve is to get tensorflow with GPU support running on a MapR (one of the many Hadoop Distributions) Cluster in Docker containers in order to analyze and explore data in a bigger scale. The GPU nodes are part of the cluster.
Therefore MapR provides a docker [data-science-refinery](https://hub.docker.com/r/maprtech/data-science-refinery/) container. This container is with 3 Gigabyte already quite big and provides a very nice package of data science/data analytics tools like Hadoop Client, Apache Pig, Spark, Drill, Zeppelin and the MapR client is preconfigured for the usage in a cluster. 
This means I would like to create a container based on the data-science-refinery container to save the work to install all binaries by myself. I expected that it would be less work to bring the cuda driver to the MapR container rather than bringing a massive workbench to a cuda container.
What would be your approach to accomplish that?

**Comment by flx42** [comment]
Don't use a multi-stage build, look at our `9.0-base` Dockerfile to see how to setup the repo, then `apt-get install` the dependencies you need, including cuDNN.
